src/transformers/models/modernbert/modular_modernbert.py

```diff
@@ -0,0 +1,1231 @@
+        Flash Attention implementation does not. If using SDPA we first call `pad_input`. Once we compute
+        attention, we re-unpad our outputs for the other layers. The pad/unpad operations add overhead, but not
+        sending pad tokens through ffs saves compute.
+
+        Args:
+            hidden_states: (total_nnz, dim)
+            cu_seqlens: (batch + 1,)
+            max_seqlen: int
+            indices: (total_nnz,)
+            attn_mask: (batch, max_seqlen)
+
+        Returns:
+            attention: (total_nnz, dim)
+        """
+        bs, dim = hidden_states.shape[:2]
+        qkv = self.Wqkv(hidden_states)
+        qkv = qkv.view(-1, 3, self.num_heads, self.head_dim)
+
+        attn = MODERNBERT_ATTENTION_FUNCTION[self.config._attn_implementation](
+            self,
+            qkv=qkv,
+            rotary_emb=self.rotary_emb,
+            position_ids=position_ids,
+            attention_mask=attention_mask,
+            local_attention=self.local_attention,
+            cu_seqlens=cu_seqlens,
+            max_seqlen=max_seqlen,
+            bs=bs,
+            dim=dim,
+        )
+
+        return self.out_drop(self.Wo(attn))
+
+
+class ModernBertEncoderLayer(nn.Module):
+    def __init__(self, config: ModernBertConfig, layer_id: Optional[int] = None):
+        super().__init__()
+        self.config = config
+        if config.skip_first_prenorm and config.embedding_norm and layer_id == 0:
+            self.attn_norm = nn.Identity()
+        else:
+            self.attn_norm = nn.LayerNorm(config.hidden_size, eps=config.norm_eps, bias=config.norm_bias)
+        self.attn = ModernBertAttention(config=config, layer_id=layer_id)
+        self.mlp_norm = nn.LayerNorm(config.hidden_size, eps=config.norm_eps, bias=config.norm_bias)
+        self.mlp = ModernBertMLP(config)
+
+    def _init_weights(self, reset_params: bool = False):
+        if reset_params:
+            self.attn_norm.reset_parameters()
+            self.mlp_norm.reset_parameters()
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        position_ids: Optional[torch.LongTensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        cu_seqlens: Optional[torch.Tensor] = None,
+        max_seqlen: Optional[int] = None,
+    ) -> torch.Tensor:
+        """Forward pass for a ModernBert layer, including both attention and MLP.
+
+        Args:
+            hidden_states: (total_nnz, dim)
+            position_ids: (total_nnz,)
+            attention_mask: (batch, max_seqlen)
+            cu_seqlens: (batch + 1,)
+            max_seqlen: int
+        """
+        attn_out = hidden_states + self.attn(
+            self.attn_norm(hidden_states),
+            position_ids=position_ids,
+            cu_seqlens=cu_seqlens,
+            max_seqlen=max_seqlen,
+            attention_mask=attention_mask,
+        )
+        return attn_out + self.mlp(self.mlp_norm(attn_out))
```

Comment:
```suggestion

    @torch.compile(dynamic=True)
    def compiled_mlp(self, hidden_states: torch.Tensor) -> torch.Tensor:
        return self.mlp(self.mlp_norm(hidden_states))

    def forward(
        self,
        hidden_states: torch.Tensor,
        position_ids: Optional[torch.LongTensor] = None,
        attention_mask: Optional[torch.Tensor] = None,
        cu_seqlens: Optional[torch.Tensor] = None,
        max_seqlen: Optional[int] = None,
    ) -> torch.Tensor:
        """Forward pass for a ModernBert layer, including both attention and MLP.

        Args:
            hidden_states: (total_nnz, dim)
            position_ids: (total_nnz,)
            attention_mask: (batch, max_seqlen)
            cu_seqlens: (batch + 1,)
            max_seqlen: int
        """
        attn_out = hidden_states + self.attn(
            self.attn_norm(hidden_states),
            position_ids=position_ids,
            cu_seqlens=cu_seqlens,
            max_seqlen=max_seqlen,
            attention_mask=attention_mask,
        )
        return attn_out + self.compiled_mlp(attn_out)
```
Required to match research implementation's performance. With this in place, the difference between the original outputs and ModernBert become:
```
tensor([[[0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         ...,
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.]]], device='cuda:0',
       dtype=torch.bfloat16, grad_fn=<SubBackward0>)
Min: 0.0
Max: 0.0
Mean: 0.0
Std: 0.0
```